### DISCUSSION TITLE
Macro to Toggle Half and full width characters

### DISCUSSION THREAD

**Discussion by Here2Fore** [texstudio-org/texstudio#1577] [Q&A]
I wonder whether it's possible to make a macro script to change half-width Unicode characters into full-width (and the reverse). This affects the original lower ASCII characters (empty space, 0..9, a..z, A..Z and punctuation), which are all half-width when used in European languages, but which  when part of Kanji/Hanzi text must be treated as full-width characters, see https://www.unicode.org/charts/PDF/UFF00.pdf (and https://www.unicode.org/charts/PDF/U3000.pdf), for in order to achieve an eye-pleasing result, all full-width characters including the empty space must be rendered with the CJKmainfont setting of the document. The problem is that when copying and pasting and/or typing in Chinese or Japanese text, the "lower ASCII" characters often default to half-width. Conversely, when full-width numerals and punctuation slip into normal half-width text, the results are equally awful. A macro (like the one to change lower to UPPER case) would be very convenient for cleaning up text.  I can't get my head around the macro-scripting, so hope that somebody is able to help.

**Comment by Here2Fore** [comment]
I hope this is the right place to ask the above question. Just to be more specific, I have no experience with Javascript or how it can be used in txs macros, so I am looking for a way to make this macro. To give a clearer impression of it, I have indicated below a possible substitution procedure:

(a) for i='\u0021'; i<='\u007E'; i++ do replace with '\uFF01'  (until \uFF5E) --- This involves punctuation and latin alfanumerics 

(b) for i='\u0020' do replace with '\u3000' --- This is the empty space 

(c) for \u00A2 , \u00A3, \u00A5, \u20A9 do replace with \uFFE0, \uFFE1, \uFFE5, \uFFE6 --- These are currency symbols in addition to $ which is included in the fisrt range

All other code points relate to (Japanese) katakana and (Korean) hangul characters and some rarely used CJK characters for which there are no relevant equivalents in Latin scripts. People may find use for a similar macro for in particular the Japanese and Korean-related characters, but I do not have the knowledge to evaluate it. 

In short, the task would be to create two macros for points a-c above, one from fullwidth to halfwidth, the other from halfwidth to fullwidth.

All the best.

**Comment by sunderme** [comment]
since you already started with pseudo code, here is a macro which goes through every text line.
You "just" need to adapt your code to change the text.
https://github.com/texstudio-org/texstudio/wiki/Scripts#remove-whitespace-at-the-end-of-all-lines

**Reply by Here2Fore** [reply]
Thank you so much for the advice, Being a total newcomer to java-scripting, I have come up against a several issues. To get the iteration going ( with i++), I found that codePointAt apparently allows us to handle strings as integers, so I wonder if defining the halfwidth (hw) and the fullwidth (fw) characters as linked (by adding the numerical difference (\uFEE0) to  the hw value) is the proper way of achieving this. It's also not clear to me whether I have got the for loop and substitution right. Furthermore, I am not sure how to bind the function to selection or full text of the document. Actually, it is crucial to select quite precisely because it would be pretty awful to change LaTeX markup into fw characters!

It will not come as a surprise that my following attempt does not work. Any further suggestions?

%SCRIPT
var t = cursor.hasSelection()?cursor.selectedText():editor.text();
var hw = string.codePointAt([\u0021]);
var fw = string.codePointAt([hw+\uFEE0]);
for (var i=hw;i<=string.codePointAt([\u007E]);i++) 
fw[i] = fw[i].replace(/hw/fw);

**Reply by sunderme** [reply]
fw needs to be updated inside the loop and you want to replace the charcters in the string "t"
and finally you need to write the resuklt back into the editor, i.e. cursor.insertText(t)

**Reply by Here2Fore** [reply]
Again thank you so much for the good advice. I am still trrying to stitch it together in the right way, but only get "parse error" as an outcome.

I have now arrived at this:

%SCRIPT
t = cursor.hasSelection()?cursor.selectedText():editor.text();
s = t.codePoints().forEach( 
var hw = s.codePointAt([\u0021]);
var fw = s.codePointAt([hw+\uFEE0]);
for (var i=hw;i<=s.codePointAt([\u007E]);i++){
for (var i=fw;i<=s.codePointAt([\uFF5E]);i++){
fw[i] = fw[i].replace(/hw/fw) } } );
cursor.insertText(t);

So I pick up the highlighted text (t), chop it up into utf-8 codepoints, each of which is a sign (s), and start a for loop for every sign in the selection. In the for loop, I define the halfwidth (hw) and fullwidth (fw) signs (s) as \u0021 (0) and \uFEED (０) respectively and in a dual for loop replace all the hw with fw up till \uFF5E (～) and end up by reinserting the text to cursor after the loops have been concluded. I cannot figure out where the parse error is, and I am not clear exactly whether this will allow for the replacement to be carried into the "insert" on the last line. At least I hope I am on the right track.

**Comment by sunderme** [comment]
I assume, you want something like this:
```
%SCRIPT
t = cursor.selectedText()
var result = new String()
for (i = 0; i < t.length; i++) { 
  var n = t.charCodeAt(i);
  if(n>=33 && n<=126){
	n=n+65248
  }
  result=result+String.fromCharCode(n);
}
cursor.insertText(result);
```

**Reply by Here2Fore** [reply]
This works elegantly. Thank you so much. I'll try to use your code as a basis for also including the further few relevant characters (mainly the full-width space, two more Chinese half-width punctuation signs for which there exists full-width forms, plus the currency symbols) and to create a separate macro with the reverse function, which is also very useful. When I have something that works, I'll be back.

**Comment by Here2Fore** [comment]
This is the result, and it works for me. Hope it conforms to txs norms. As it may be of use to others, I suggest that it be made available. Again, thanks

**CJK Full- to Halfwidth Transformation**
Ref.: Unicode _CJK symbols and punctuation_ (3000-303F) and _Halfwidth and fullwidth forms_ (FF00-FFEF)

%SCRIPT
// Script to change Unicode fullwidth character forms \uFF01-\uFF5E, 
// \u3000-\u3002, \uFFE0-\uFFE1, and \uFFE5-\uFFE6 to equivalent halfwidth 
// forms (半角形式). The script also transforms CJK punctuation characters 
// (、and 。) as well as currency symbols, but omits CJK brackets plus 
// Katakana and Hangul characters. The script helps "clean" CJK text 
// snippets to facilitate inline use in alphabetic text.
//  
t = cursor.selectedText()
var result = new String()
for (i = 0; i < t.length; i++) { 
  var n = t.charCodeAt(i);
if (n>=65281 && n<=65374){	
  	n = n-65248;
  	}
else
if (n == 12288){ 
 	n = 32;
 	}
else
if  (n == 12289){
	n = 65380;
	}
else
if  (n == 12290){
  	n = 65377;
  	}
else
if  (n == 65504){
 	n = 162;
 	}
else
if (n == 65505){
 	n = 163;
 	}
else
if  (n == 65509){
 	n = 165;
}
else
if  (n == 65510){
 	n = 8361;
 	}
else
	{
	n = n;
	}
result=result+String.fromCharCode(n);
}
cursor.insertText(result);


**CJK Half- to Fullwidth Transformation**
Ref.: Unicode _CJK symbols and punctuation_ (3000-303F) and _Halfwidth and fullwidth forms_ (FF00-FFEF)


%SCRIPT
// Script to change Unicode halfwidth character forms \u0020-\u007E, 
// \u0020,\u002C, \u002E, \u00A2-\u00A3, \u00A5, and \u00A9 to equivalent fullwidth 
// forms. The script also transforms CJK punctuation characters 
// (､and ｡) as well as currency symbols, but omits CJK brackets plus 
// Katakana and Hangul characters. The script helps "clean" CJK text 
// blocks to ensure coherent effect of CJK fonts.
// 
t = cursor.selectedText()
var result = new String()
for (i = 0; i < t.length; i++) { 
  var n = t.charCodeAt(i);
  if(n>=33 && n<=126){
	n=n+65248
  }
  else
if (n == 32){ 
 	n = 12288;
 	}
else
if  (n == 12289){
	n = 65380;
	}
else
if  (n == 12290{
  	n = 65377);
  	}
else
if  (n == 162){
 	n = 65504;
 	}
else
if (n == 163){
 	n = 65505;
 	}
else
if  (n == 165){
 	n = 65509;
}
else
if  (n == 8361){
 	n = 65510;
 	}
else
	{
	n = n;
	}
result=result+String.fromCharCode(n);
}
cursor.insertText(result);
